### PR TITLE
Add ESP32 support

### PR DESCRIPTION
**Pull request by silps**
I only tested the websocket server on my esp32 and it worked fine
I use platformio and looks like the esp32 arduino libs are not up to date, not the ones on github
so I had lots of trouble finding the right functions, may be it is better to use the up to date ones

### CODE REVIEW

**Comment by SamuelM333** [comment]
Do you have a working example for ESP32? Also, which other dependencies do I need to build this? I can't compile a PlatformIO app with `WebSockets.h` or any header from this plugin imported. This is the output:

```
D:\Development\Git\serverchip\serverchip-esp32>pio run
[08/25/17 19:26:20] Processing esp32dev (platform: espressif32; lib_deps: ArduinoJson, https://github.com/Links2004/arduinoWebSockets/archive/master.zip; board: esp32dev; framework: arduino)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
Collected 17 compatible libraries
Looking for dependencies...
Library Dependency Graph
|-- <ArduinoJson> v5.11.1
|-- <HTTPClient>
|   |-- <WiFi> v1.0
|-- <WiFi> v1.0
|-- <WebSockets> v2.0.10
|   |-- <SPI> v1.0
Compiling .pioenvs\esp32dev\src\main.o
In file included from src\main.cpp:4:0:
.piolibdeps\WebSockets\src/WebSockets.h:131:22: fatal error: Ethernet.h: No such file or directory
compilation terminated.
Compiling .pioenvs\esp32dev\lib\SPI\SPI.o
Compiling .pioenvs\esp32dev\lib\WebSockets\WebSockets.o
*** [.pioenvs\esp32dev\src\main.o] Error 1
Compiling .pioenvs\esp32dev\lib\WebSockets\WebSocketsClient.o
In file included from .piolibdeps\WebSockets\src\WebSockets.cpp:25:0:
.piolibdeps\WebSockets\src\WebSockets.h:131:22: fatal error: Ethernet.h: No such file or directory
compilation terminated.
In file included from .piolibdeps\WebSockets\src\WebSocketsClient.cpp:25:0:
.piolibdeps\WebSockets\src\WebSockets.h:131:22: fatal error: Ethernet.h: No such file or directory
compilation terminated.
*** [.pioenvs\esp32dev\lib\WebSockets\WebSockets.o] Error 1
*** [.pioenvs\esp32dev\lib\WebSockets\WebSocketsClient.o] Error 1
```

I can't figure out which `Ethernet.h` file is trying to import. I tried with [Ethernet 134](http://platformio.org/lib/show/134/Ethernet) and [Ethernet 872](http://platformio.org/lib/show/872/Ethernet) with no luck, it just spits more errors now about `Ethernet.h` (the library with ID 134 gave me the least amount of errors).

This is my `platformio.ini`:

```
[env:esp32dev]
platform = espressif32
board = esp32dev
framework = arduino
lib_deps =
    ArduinoJson
    https://github.com/Links2004/arduinoWebSockets/archive/master.zip ; Using master branch because this pull request isn't on a release version yet
```

**Comment by silps** [comment]
That's strange, it shouldn't include Ethernet.h if you compile for ESP32. The ESP32 should be defined and it should use
#include <WiFi.h>
#include <WiFiClientSecure.h>

**Comment by SamuelM333** [comment]
Ok now I'm using [this](https://github.com/Links2004/arduinoWebSockets/tree/4fc80871a605694bf09f3c0503fabd43e3bd2150) version of the plugin with the changes that you introduced and it manages to build now, but I have to use espressif32_stage  on my `platformio.ini`, probably because some libs are outdates as you said at the start of this PR. 

Now I'm trying to run an example, this one would be great [WebSocketClientSocketIO.ino](https://github.com/Links2004/arduinoWebSockets/blob/4fc80871a605694bf09f3c0503fabd43e3bd2150/examples/WebSocketClientSocketIO/WebSocketClientSocketIO.ino). After making the necessary import changes it manages to build, but I have to comment out line 55 becase it doesn't find `hexdump`. Is that important? How do I fix that?

Also, I flashed the program with that line commented out for testing but now my ESP32 doesn't boot. I guess it is a problem with the stage branch of the ESP32 Arduino core, because when I run some other example projects with the stable build, everything goes fine.

This is the output when I try to build the project with the stable branch:

```
D:\Development\Git\serverchip\serverchip-esp32>pio run
[08/26/17 15:58:22] Processing esp32dev (platform: espressif32; lib_deps: ArduinoJson; board: esp32dev; framework: arduino)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
Collected 16 compatible libraries
Looking for dependencies...
Library Dependency Graph
|-- <ArduinoJson> v5.11.1
|-- <WiFi> v1.0
|-- <WebSockets> v2.0.7
|   |-- <WiFi> v1.0
|   |-- <HTTPClient>
|   |   |-- <WiFi> v1.0
|   |-- <SPI> v1.0
Linking .pioenvs\esp32dev\firmware.elf
C:\Users\samue\.platformio\packages\framework-arduinoespressif32\tools\sdk\lib\libwpa_supplicant.a(sha1-internal.o): In function `SHA1Transform':
/Users/ficeto/Desktop/ESP32/ESP32/esp-idf-public/components/wpa_supplicant/src/crypto/sha1-internal.c:180: multiple definition of `SHA1Transform'
.pioenvs\esp32dev\lib\libWebSockets.a(libsha1.o):D:\Development\Git\serverchip\serverchip-esp32/lib\WebSockets\src\libsha1/libsha1.c:58: first defined here
C:\Users\samue\.platformio\packages\framework-arduinoespressif32\tools\sdk\lib\libwpa_supplicant.a(sha1-internal.o): In function `SHA1Init':
sha1-internal.c:(.text.SHA1Init+0x0): multiple definition of `SHA1Init'
.pioenvs\esp32dev\lib\libWebSockets.a(libsha1.o):libsha1.c:(.text.SHA1Init+0x0): first defined here
C:\Users\samue\.platformio\packages\framework-arduinoespressif32\tools\sdk\lib\libwpa_supplicant.a(sha1-internal.o): In function `SHA1Update':
sha1-internal.c:(.text.SHA1Update+0x0): multiple definition of `SHA1Update'
.pioenvs\esp32dev\lib\libWebSockets.a(libsha1.o):libsha1.c:(.text.SHA1Update+0x0): first defined here
C:\Users\samue\.platformio\packages\framework-arduinoespressif32\tools\sdk\lib\libwpa_supplicant.a(sha1-internal.o): In function `SHA1Final':
sha1-internal.c:(.text.SHA1Final+0x0): multiple definition of `SHA1Final'
.pioenvs\esp32dev\lib\libWebSockets.a(libsha1.o):libsha1.c:(.text.SHA1Final+0x0): first defined here
collect2.exe: error: ld returned 1 exit status
*** [.pioenvs\esp32dev\firmware.elf] Error 1
```

Basically, I'm stuck. If I use the stable build I can't compile the project. If I use the stage branch, my ESP32 doesn't boot...

**Comment by silps** [comment]
I see, I remember getting into the same problem with the SHA functions. I have to check what I did to fix it.

**Comment by silps** [comment]
Technically it shouldn't try to include the `libsha1.h`, hmm.

#ifdef ESP8266
#include <Hash.h>
#elif defined(ESP32)
#include <hwcrypto/sha.h>
#else

extern "C" {
#include "libsha1/libsha1.h"
}

#endif

**Comment by SamuelM333** [comment]
Yeah sorry, I was using a version of the plugin that didn't have that statement. Now I'm using the [master](https://github.com/robokoding/arduinoWebSockets) branch of your fork and it builds (I had to delete a trailing `}` on the file `WebSocketsServer.cpp` at line 622).

Now I'm trying to use the SocketIO example with no luck. My SocketIO server acknowledges a connection but it doesn't trigger the `connected` event. On my ESP32 logs I have this: 

```
[WSc] Connected to url: /socket.io/?EIO=3
[WSc] Disconnected!
[WSc] Connected to url: /socket.io/?EIO=3
Sending msg... // I added this print inside loop() just before the webSocket.sendTXT call to check if it was happening
[E][WiFiClient.cpp:234] read(): 104
[WSc] Disconnected!
[WSc] Connected to url: /socket.io/?EIO=3
... // This pattern repeats  
```

Apparently, it's stuck between getting connected and disconnected. And when I manages to send the message, it triggers an exception on `WiFiClient.cpp`

Any idea? Thanks for your help.

**Comment by silps** [comment]
I see, I don't have many thoughts currently. I tested WebSocketClient against a nginx server and it was able to send and receive messages. May be also try the WebSocketServer and see if that works.

**Comment by tetious** [comment]
The extra close bracket on 622 is actually not extra, but due to line 590 missing ESP32 from the if test. Without this, the code goes into a loop thinking a connection was made when it actually wasn't. :)

Line 590 should look like this:
`
#if (WEBSOCKETS_NETWORK_TYPE == NETWORK_ESP8266) || WEBSOCKETS_NETWORK_TYPE == NETWORK_ESP32)
`

**Comment by silps** [comment]
@tetious thanks for the tip. @SamuelM333 I added updates now and tested the client again. It worked for me screen shot 2017-09-03 at 10 21 19 am

**Comment by SamuelM333** [comment]
Thanks for all your help and efforts. I still have problems compiling the lib. Again, I'm using the [master](https://github.com/robokoding/arduinoWebSockets) branch of @silps fork
```
Compiling .pioenvs\esp32dev\lib\WebSockets\WebSocketsServer.o
lib\WebSockets\src\WebSocketsServer.cpp: In member function 'void WebSocketsServer::handleNewClients()':
lib\WebSockets\src\WebSocketsServer.cpp:591:20: error: 'class WiFiServer' has no member named 'hasClient'
while(_server->hasClient()) {
^
*** [.pioenvs\esp32dev\lib\WebSockets\WebSocketsServer.o] Error 1
```

I ordered an ESP8266 (ESP-12E) to overcome these issues, I really need to get a project going soon. I'll wait for a while until there's more support for the ESP32. Apparently, everything works for the ESP8266.

**Comment by tetious** [comment]
I'm actively using the server with my fork, which isn't that different from @silps fork.

Maybe try it and see if you have any luck? If it works, maybe I can figure out how to add my tweaks to this PR. ;)

**Comment by tetious** [comment]
The issue you mentioned earlier with disconnects might be related to the bug I just fixed where it was using the AVR path when defining max payload size. 

The code disconnects without a word if you try to send too much, and the AVR path is set to 1K (!!).

Dunno how it will hold up long term, but I set the ESP32 to use 32K.

**Comment by SamuelM333** [comment]
@tetious Your fork gives me the same error, `lib\WebSockets\src\WebSocketsServer.cpp:591:20: error: 'class WiFiServer' has no member named 'hasClient'`

Are you using PlatformIO?

**Comment by tetious** [comment]
Yup. Have you updated your toolchain recently? (via http://docs.platformio.org/en/latest/userguide/cmd_update.html)

**Comment by SamuelM333** [comment]
Ah damn, indeed I had a very old version of the espressif32 (0.7.5  when the up to date is 0.9.1). Thank you for that, I thought PlatformIO would take care of this.

It compiles now. I tried the SocketIO example but it falls into the loop you mentioned. My WebSocket server starts to spit a lot of logs with connections (I'm using a Python app with Flask and Flask SocketIO):

```
73096) accepted ('192.168.1.119', 3585)
192.168.1.119 - - [03/Sep/2017 15:36:05] "GET /socket.io/?EIO=3&transport=polling HTTP/1.1" 200 387 0.000000
192.168.1.119 - - [03/Sep/2017 15:36:05] "GET /socket.io/?EIO=3&transport=websocket&sid=474bc840564f457f85535b2ee3399a36 HTTP/1.1" 200 0 0.028000
```
And on my ESP32:
```
[WSc] Connected to url: /socket.io/?EIO=3
[WSc] Disconnected!
[WSc] Connected to url: /socket.io/?EIO=3
[WSc] Disconnected!
```

(These outputs repeat)

And if I try the WebSocketClient example, this is the output on my server
```
192.168.1.119 - - [03/Sep/2017 15:29:16] "GET / HTTP/1.1" 200 552 0.001000
192.168.1.119 - - [03/Sep/2017 15:29:16] code 400, message Bad request syntax ('This is a webSocket client!')
192.168.1.119 - - [03/Sep/2017 15:29:16] "This is a webSocket client!" 400 -
```
And on the ESP32:
`[WSc] Disconnected!`

**Comment by Links2004** [comment]
Have created a new branch for the ESP32 implementation, 
as soon all is working it can be integrated in master.

**Comment by SamuelM333** [comment]
@Links2004 Awesome, thanks. I can test the ESP32 implementation whenever it's ready. So far I haven't been able to use it. I had to fall back to an ESP8266 to keep working on my project. I'm using [this](https://github.com/timum-viw/socket.io-client) plugin as a wrapper and and it works well.

**Comment by tetious** [comment]
I haven't had a chance to dig further, but I think the problem with the client support is that none of the include guards deal with ESP32 presently.

I started working on generalizing the guards a bit, as the ESP* works the same nearly everywhere. The biggest difference is some of the include files. 

I don't have time right now to test and create a PR, but if you want to try the **untested** (it seems to compile at least) code in my fork (https://github.com/tetious/arduinoWebSockets) maybe it will work for you? :)

If it seems sensible and tests ok, I'll do a PR against the new branch.
